src/FSharp.Data.GraphQL.Shared/Serialization.fs

```diff
@@ -0,0 +1,168 @@
+      let extractors =
+        [
+          for pi in FSharpType.GetRecordFields(typ) do
+            let k = pi.Name
+            let v = generateSerializerInteral pi.PropertyType
+
+            fun t ->
+              k, v (pi.GetValue(t))
+        ]
+
+      fun (t : obj) ->
+        [
+          for extract in extractors do
+            extract t
+        ]
+        |> Map.ofSeq
+        |> ObjectValue
+    else if FSharpType.IsUnion(typ, true) then
+      fun (v : obj) ->
+        let info, _ = FSharpValue.GetUnionFields(v, typ, true)
+        StringValue info.Name
+    else if FSharpType.IsTuple(typ) then
+      let serializers =
+        FSharpType.GetTupleElements(typ)
+        |> Array.map generateSerializerInteral
+
+      fun (v : obj) ->
+        FSharpValue.GetTupleFields(v)
+        |> Seq.map2 (fun serializer x -> serializer x) serializers
+        |> Seq.toList
+        |> ListValue
+    else if (
+      typ.IsGenericType
+      && typ.GetGenericTypeDefinition().FullName = typedefof<Map<IComparable, obj>>.FullName
+      && typ.GenericTypeArguments[0].FullName = typeof<string>.FullName) then
+      let keyType = typ.GenericTypeArguments[0]
+      let valueType = typ.GenericTypeArguments[1]
+      let valueEncoder = generateSerializerInteral valueType
+      let kvProps = typedefof<KeyValuePair<obj,obj>>.MakeGenericType(keyType, valueType).GetProperties()
+
+      let valueSerializer = generateSerializerInteral typ.GenericTypeArguments[1]
+
+      fun (v : obj) ->
+        [
+          for kvp in v :?> System.Collections.IEnumerable do
+            let k = kvProps[0].GetValue(kvp) :?> string
+            let v = kvProps[1].GetValue(kvp)
+
+            k, valueSerializer v
+        ]
+        |> Map.ofSeq
+        |> ObjectValue
+    else if typ.IsEnum then
+      let enumType = Enum.GetUnderlyingType(typ).FullName
+
+      if enumType = typeof<sbyte>.FullName then
+        fun (v : obj) -> IntValue (box v :?> sbyte |> int64)
+      elif enumType = typeof<byte>.FullName then
+        fun (v : obj) -> IntValue (box v :?> byte |> int64)
+      elif enumType = typeof<int16>.FullName then
+        fun (v : obj) -> IntValue (box v :?> int16 |> int64)
+      elif enumType = typeof<uint16>.FullName then
+        fun (v : obj) -> IntValue (box v :?> uint16 |> int64)
+      elif enumType = typeof<int>.FullName then
+        fun (v : obj) -> IntValue (box v :?> int |> int64)
+      elif enumType = typeof<uint32>.FullName then
+        fun (v : obj) -> IntValue (box v :?> uint32 |> int64)
+      else
+        failwith $"Unsupported underlying enum type {enumType}"
+    else
+      // This behaviour is likely not what the user wants.
+      // However, if they are not using standard F# constructs then
+      // a custom serializer is the best option.
+      fun x -> StringValue (string x)
+
+  let generateSerializer<'t> () : ('t -> Value) =
+    fun (t : 't) ->
+      generateSerializerInteral (typeof<'t>) (box t)
```

Comment:
@njlr can a piece of this be replaced by `FSharp.SystemTextJson`?

Reply:
I don't think so - this code generates a function from `'t -> FSharp.Data.GraphQL.Ast.Value` rather than `'t -> string`